client/src/components/Draftail/CommentableEditor/CommentableEditor.js

```diff
@@ -211,14 +222,37 @@ function CommentableEditor({commentApp, fieldNode, contentPath, rawContentState,
       // This hack can be removed when draft-js triggers inline style rerender on props change
       setEditorState(editorState => forceResetEditorState(editorState, 
         Modifier.applyInlineStyle(
-          filteredContent,
+          filterInlineStyles(inlineStyles.map(style => style.type).concat(ids.map(id => COMMENT_STYLE_IDENTIFIER + id)), editorState.getCurrentContent()),
           getFullSelectionState(filteredContent),
           'STYLE_RERENDER_'+uniqueStyleId
         )
       ))
       setUniqueStyleId((id) => (id + 1) % 200);
     }, [focusedId, enabled, inlineStyles, ids, editorState])
 
+    useEffect(() => {
+      // if there are any comments without annotations, we need to add them to the EditorState
+      let contentState = editorState.getCurrentContent();
+      let hasUpdated = false;
+      comments.filter(comment => !comment.annotation).forEach((comment) => {
+        commentApp.updateAnnotation(new DraftailInlineAnnotation(fieldNode), comment.localId);
+        const style = `${COMMENT_STYLE_IDENTIFIER}${comment.localId}`
+        try {
+          const position = JSON.parse(comment.position)
```

Comment:
My logic for doing it this way was:
- Actually having this as an object is really only necessary when loading new comments (for Draftail, anyway)
- When updating comment positions, we pretty much just want to overwrite the old comment position with the new one, so a stringified version also makes sense
- wagtail-comment-frontend leaves it to the widget to deal with the position, it doesn't actually do anything with it itself other than output it to the form (and that of course needs the string)

If we have custom widgets in the future that are doing a lot of work with the deserialized position, or updating it a lot, I can see the object version being better though. (If we do have it deserialised in wagtail-comment-frontend though, maybe it would be time to refactor to use something like `immer` to avoid any deepcopying headaches?)

Reply:
I think your points are valid, but the main reason I find this odd nonetheless is that we tend to prefer separating out the "business logic" code with the "data storage" code, regardless of the actual use of the data.

The one thing that wagtail-comment-frontend does with the position is to store it – so I think it would be more appropriate for it to deal with any storage requirements such as serialisation. For example, if comments gained more features requiring more fields in the future, I can see how we might want to serialize the whole comment in a single field as JSON, rather than having different serialisation logic for each field – in which case it might be much simpler to already have said logic all in the `wagtail-comment-frontend` code, rather than there + here and there in business logic.

Reply:
Thanks! As you suggested it's not a big issue, I'm going to merge this now, and hopefully make this serialization change soon, now I've been refactoring the reducer to use immer, which should reduce the burden of storing arbitrary deserialized json in the store.